workspace/__lib__/xod/waves/square-wave/patch.cpp

```diff
@@ -0,0 +1,53 @@
+struct State {
+    bool wasEnabled;
+    TimeMs timeToSwitch;
+    TimeMs nextSwitchTime;
+};
+
+{{ GENERATED_CODE }}
+
+void evaluate(Context ctx) {
+    State* state = getState(ctx);
+    TimeMs t = transactionTime();
+
+    bool enabled = getValue<input_EN>(ctx);
+    bool reset = isInputDirty<input_RST>(ctx);
+    Number period = getValue<input_T>(ctx);
+    Number duty = getValue<input_DUTY>(ctx);
+
+    if (reset) {
+        emitValue<output_OUT>(ctx, enabled);
+        emitValue<output_N>(ctx, 0);
+        clearTimeout(ctx);
+        // enforce rescheduling at the next stage if enabled
+        state->wasEnabled = false;
+    }
+
+    if (enabled && !state->wasEnabled) {
+        // just enabled/resumed
+        state->timeToSwitch = (period * duty) * 1000.0;
+        setTimeout(ctx, state->timeToSwitch);
+        state->nextSwitchTime = t + state->timeToSwitch;
+        emitValue<output_OUT>(ctx, true);
+    } else if (!enabled && state->wasEnabled) {
+        // just paused
+        // TODO: we can get rid of storing nextSwitchTime if API would
```

Comment:
This particular implementation is just a copy of `xod/core/square-wave`.
I'm not sure about growing the API for a single use-case. Do you have some examples where it could also be useful?

Reply:
Suggest leaving it as is for now to limit the PR scope